testsuite/integration/elytron/src/test/java/org/wildfly/test/integration/elytron/realm/DistributedRealmTestCase.java

```diff
@@ -0,0 +1,346 @@
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.wildfly.test.integration.elytron.realm;
+
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URISyntaxException;
+import static javax.servlet.http.HttpServletResponse.SC_OK;
+import static org.junit.Assert.assertEquals;
+
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.Base64.Encoder;
+import static javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR;
+import static javax.servlet.http.HttpServletResponse.SC_UNAUTHORIZED;
+import org.apache.http.Header;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.ClientProtocolException;
+import org.apache.http.client.methods.HttpGet;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClientBuilder;
+import org.apache.http.util.EntityUtils;
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.arquillian.container.test.api.OperateOnDeployment;
+import org.jboss.arquillian.container.test.api.RunAsClient;
+import org.jboss.arquillian.junit.Arquillian;
+import org.jboss.arquillian.test.api.ArquillianResource;
+import org.jboss.as.arquillian.api.ServerSetup;
+import org.jboss.as.test.integration.security.common.Utils;
+import static org.jboss.as.test.integration.security.common.Utils.REDIRECT_STRATEGY;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.StringAsset;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assert.assertTrue;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.wildfly.test.security.common.AbstractElytronSetupTask;
+import org.wildfly.test.security.common.elytron.ConfigurableElement;
+import org.wildfly.test.security.common.elytron.DirContext;
+import org.wildfly.test.security.common.elytron.DistributedRealm;
+import org.wildfly.test.security.common.elytron.LdapRealm;
+import org.wildfly.test.security.common.elytron.MechanismConfiguration;
+import org.wildfly.test.security.common.elytron.PropertiesRealm;
+import org.wildfly.test.security.common.elytron.SimpleHttpAuthenticationFactory;
+import org.wildfly.test.security.common.elytron.SimpleSecurityDomain;
+import org.wildfly.test.security.common.elytron.TokenRealm;
+import org.wildfly.test.security.common.elytron.UserWithAttributeValues;
+import org.wildfly.test.undertow.common.UndertowApplicationSecurityDomain;
+
+/**
+ * Tests the {@link DistributedRealm} within the Elytron subsystem.
+ *
+ * @author Ondrej Kotek
+ */
+@RunWith(Arquillian.class)
+@RunAsClient
+@ServerSetup({DistributedRealmTestCase.ServerSetup.class})
+public class DistributedRealmTestCase {
```

Comment:
Other tests in the project are not a good example to go by.

One issue I have is in two years time something may happen and suddenly we have a failing test in CI, where a test relies on multiple resources potentially "engineered" to behave in a certain way it can often be time consuming to try and reverse engineer the original intent of the test before getting to the failure itself.

A javadoc comment describing the general relationship of resources under test would certainly help a lot.

I am not marking this as requesting changes but if you did have time that would be great.

Reply:
I added the general javadoc comment and several comments to the code to make it clear. I appreciate further feedback.